docs/RFCS/index_hints.md

```diff
@@ -0,0 +1,125 @@
+- Feature Name: index_hints
+- Status: draft
+- Start Date: 2014-03-31
+- Authors: Radu
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #5625
+
+# Summary
+
+The proposal is to add syntax to allow the user to force using a specific index
+for a query. This is intended as a workaround when the index selection algorithm
+results in a bad choice.
+
+# Motivation
+
+The index selection algorithm takes into account a lot of factors but is far
+from perfect. One example we saw recently was from the photos app, where a
+`LIMIT` clause would have made using an index a much better choice. We have
+fixed that problem since, but there are other known issues, e.g. [#5589][5589].
+
+When we wanted a quick workaround for the photos app issue, we had to resort to
+using a crafted subquery.
+
+We want to provide an easy way to work around this kind of problem.
+
+[5589]: https://github.com/cockroachdb/cockroach/issues/5589
+
+# Current `@` syntax
+
+We currently support special syntax which allows us to use an index as a
+separate table, one which only has those columns that are stored in the index:
+
+```sql
+CREATE TABLE test (
+    k INT PRIMARY KEY,
+    u INT,
+    v INT,
+    w INT,
+    INDEX uv (u, v)
+);
+
+INSERT INTO test VALUES (1, 10, 100, 1000), (2, 20, 200, 2000);
+SELECT * FROM test@uv;
+```
+
+| u  |  v  |
+|----|-----|
+| 10 | 100 |
+| 20 | 200 |
+
+This feature might be of some use internally (for debugging) but it is not of
+much value to a user. Notably, if we had a way to force the use of a specific
+index, that syntax could be used to produce a query equivalent to the one above:
+```sql
+SELECT u,v FROM test USING INDEX uv
+```
+
+# Other systems
+
+A brief overview of what other DBMSs support:
+
+### MySQL
+
+Basic syntax:
+```sql
+-- Restricts index use to one of given indexes (or neither)
+SELECT * FROM table1 USE INDEX (col1_index,col2_index)
+
+-- Excludes some indexes from being used
+SELECT * FROM table1 IGNORE INDEX (col3_index)
+
+-- Forces use of one of the given indexes
+SELECT * FROM table1 FORCE INDEX (col1_index)
+```
+
+More syntax and detailed information [here][1].
+
+[1]: http://dev.mysql.com/doc/refman/5.7/en/index-hints.html
+
+### PostgreSQL
+
+PG does not provide support for hints. Instead they provide various knobs for
+tuning the optimizer to do the right thing. Details [here][2].
+
+[2]: http://blog.2ndquadrant.com/hinting_at_postgresql/
+
+### Oracle
+
+Oracle provides [various options][3] for hints, a basic example is:
+```sql
+SELECT /*+ INDEX(v.e2.e3 emp_job_ix) */  * FROM v
+```
+
+[3]: http://docs.oracle.com/cd/B19306_01/server.102/b14211/hintsref.htm#i26205
+
+### MS SQL Server
+
+SQL Server supports [many hints][4], the index hint is:
+
+```sql
+SELECT * FROM t WITH (INDEX(myindex))
+```
+
+[4]: https://msdn.microsoft.com/en-us/library/ms187373.aspx
+
+# Alternatives
+
+We have two options:
+
+1. Change the semantics of the existing `@` syntax
```

Comment:
After thinking about it more, this is my preference, perhaps extended to support `@{force_index=foo}` and `@{ignore_index=foo}`. I'm happy to just keep the current `@foo` syntax as well.

Given that everybody does these index hints differently and postgres doesn't support them at all, I don't see a strong reason to prefer one syntax over another.